src/org/exist/storage/XQueryPool.java

```diff
@@ -58,6 +56,8 @@
 
     private final static Logger LOG = LogManager.getLogger(XQueryPool.class);
 
+    private ConcurrentMap<Source, Deque<CompiledXQuery>> pool = new ConcurrentHashMap<>();
+
     private long lastTimeOutCheck;
     private long lastTimeOfCleanup;
 
```

Comment:
`ArrayDeque` is not thread-safe, so why you don't see problems? [javadoc](https://docs.oracle.com/javase/7/docs/api/java/util/ArrayDeque.html)

Again, I really can't understand you problem with `lastTimeOfCleanup` and `lastTimeOfCleanup`. It (mostly) always just increasing, and all check in same direction `>`.